zkvm/signature: implement MuSig protocol (#190)
interstellar/slingshot: zkvm/src/signature/signature.md

- `Multikey::aggregated_key(self)`: returns the aggregated key stored in the multikey, `self.aggregated_key`.  

### Signature

A signature is comprised of a scalar `s`, and a RistrettoPoint `R`. In the simple Schnorr signature case, `s` represents the Schnorr signature scalar and `R` represents the nonce commitment. In the MuSig signature case, `s` represents the sum of the Schnorr signature scalars of each party, or `s = sum_i (s_i)`. `R` represents the sum of the nonce commitments of each party, or `R = sum_i (R_i)`. 

Functions:
- `Signature::verify(...)`: In both the simple Schnorr signature and the MuSig signature cases, the signature verification is the same. For more detail, see the [verification](#verifying) section.

## Operations

### Key Aggregation

Key aggregation happens in the `Multikey::new(...)` function. 

Input:
- pubkeys: `Vec<VerificationKey>`. This is a list of compressed public keys that will be aggregated, as long as they can be decompressed successfully.

Operation:
- Create a new transcript using the tag "ZkVM.aggregated-key". (TODO: remove the "ZkVM." if we make the signature crate separate from the ZkVM crate.)

> - Commit all the pubkeys to the transcript. This reflects the creation of the `<L>` encoding of the list of pubkeys detailed in the MuSig paper: `<L> = H(X_1 || X_2 || ... || X_n)`. The difference is that we do not need to squeeze `<L>` out of the transcript, since we can use the prepared transcript directly for future randomness.

Reviewer: If we read the `L = H(X_1, ..., X_n)` as "commitment to all the keys", then transcript state implements exactly this, so there is no conceptual difference — the only difference is superficial — in the syntax. Because when you permute the sponge, it's _exactly just like calling a hash function_, just with a bigger output than people are used to.

```suggestion
- Commit all the pubkeys to the transcript. The transcript state corresponds to the commitment `<L>` in the MuSig paper: `<L> = H(X_1 || X_2 || ... || X_n)`.
```

Author: Yes that does sound way better, thanks!

  (TODO: this sounds awkward, explain better?)
- Create `aggregated_key = sum_i ( a_i * X_i )`. Iterate over the pubkeys, compute the factor `a_i = H(<L>, X_i)`, and add `a_i * X_i` to the aggregated key.

Output:
- a new multikey, with the transcript and aggregated key detailed above.

### Signing

There are several paths to signing:
1. Make a Schnorr signature with one public key (derived from one private key).
    Function: `Signature::sign_single(...)`

    Input: 

>     - transcript: `&Transcript` - the message to be signed should have been committed to the transcript beforehand. (TODO: pass in a mutable borrow of a transcript instead of a borrow.)

Reviewer: let's write the doc as we need this to work, and figure out borrowck issues in the implementation?
```suggestion
    - transcript: `&mut Transcript` - the message to be signed should have been committed to the transcript beforehand.
```

    - privkey: `Scalar`

    Operation:

>     - Use the transcript to generate a random factor (the nonce, `r`), by committing to the privkey and passing in a `thread_rng`

Reviewer: we need to clone transcript before mixing it with privkey and RNG.

```suggestion
    - Clone the transcript state, mix it with the privkey and system-provided RNG to generate the nonce `r`. This makes the nonce uniquely bound to a message and private key, and also makes it non-deterministic to prevent "rowhammer" attacks.
```

    - Use the nonce to create a nonce commitment `R = r * G`

>     - Mutably clone the transcript

Reviewer: We should update the user-provided transcript here.
```suggestion
```

    - Make `c = H(X, R, m)`. Because `m` has already been fed into the transcript externally, we do this by committing `X = privkey * G` to the transcript with label "P" (for pubkey), committing `R` to the transcript with label "R", and getting the challenge scalar `c` with label "c".
    - Make `s = r + c * x` where `x = privkey`

    Output:
    - Signature { s, R }

2. Make a Schnorr signature with one aggregated key (`Multikey`), derived from multiple public keys.
    TODO (tie to protocol for party & counterparty state transitions)

### Verifying

Signature verificaiton happens in the `Signature::verify(...)` function.

Input: 
- `&self`

> - transcript: `&Transcript` - the message to be signed should have been committed to the transcript beforehand. (TODO: pass in a mutable borrow of a transcript instead of a borrow.)

Reviewer: ```suggestion
- transcript: `&mut Transcript` - the message to be signed should have been committed to the transcript beforehand.
```

- P: `VerificationKey`

Operation:

> - Mutably clone the trancript 

Reviewer: ```suggestion
```

  ↓
PartyAwaitingPrecommitments{transcript, privkey, multikey, nonce, noncecommitment, Vec<Counterparty>}
  ↓
.receive_precommitments(self, Vec<precommitment>) → NonceCommitment(RistrettoPoint)
  ↓
PartyAwaitingCommitments{transcript, privkey, multikey, nonce, Vec<CounterpartyPrecommitted>}
  ↓
.receive_commitments(self, Vec<commitment>) → Share(Scalar)
  ↓
PartyAwaitingShares{multikey, c, R, Vec, CounterpartyCommitted>} 
  ↓
.receive_shares(self, Vec<share>) → Signature{s, R}

```

Note:
For now, we will have message redundancy - meaning, each party will receive and verify its own messages as well as its counterparties' messages. This makes the protocol slightly simpler, but does incur a performance overhead. (Future work: potentially remove this redundancy).

Also, for now we will assume that all of the messages passed into each party state arrive in the same order (each party's message is in the same index). This allows us to skip the step of ordering them / assigning indexes. (Future work: allow for unordered inputs, have the parties sort them.)

### Party

Fields: none

Function: `new(...)`

Input: 

> - transcript: `&Transcript` - the message to be signed should have been committed to the transcript beforehand. (TODO: pass in a mutable borrow of a transcript instead of a borrow.)

Reviewer: ```suggestion
- transcript: `&mut Transcript` - the message to be signed should have been committed to the transcript beforehand.
```

# Signatures: Engineering design doc

vanilla design:
- single message 
- no recursion
- no defense against russell's attack

2 parts of protocol:
1. key aggregation
	multikey
	keeps transcript, feeds pubkeys to transcript
	clone / fork transcript to make per-key factors
2. signing & verifying
	each party has state, wants to go to next state by verifying & performing computations. so you can "encode validity of something in the type".
	party transitions need to be non-replayable (is this true?)

dealer & parties in rangeproof:
- dealer accumulates all factors, holds onto the transcript, gives challenges
- parties don't have to know about other parties, who's failing, transcript, etc.

> oleg says: shouldn't have dealer

Reviewer: Yeah, because each party has to validate what they sign, and each other's contributions before they proceed. In RP MPC it is not that important because each party proves their own statement and dealer only stitches those together, - parties individually are not critically affected by misbehavior of others. E.g. cannot be fooled to leak the key or accidentally prove something else ("replay a signature"). Also in RP MPC we explicitly do not have a goal of verifying transcript, so parties are simply proving the RP "contextless", that is shifted to the outer protocol, which possibly would have to replay dealer's state like we would replay counterparties' states.

> create 3 counterparty structs,

Reviewer: how about this format:
```
Counterparty{pubkey}
  ↓
.precommit_nonce(H) // simply adds precommitment
  ↓
CounterpartyPrecommitted{H, pubkey}
  ↓
.commit_nonce(R) // verifies hash(R) == H
  ↓
CounterpartyCommitted{R, pubkey}
/* R_total := sum{R_i}, commit to transcript, compute and send out siglet */
  ↓
.sign(s) // verifies s_i*G == R_i + c * factor * pubkey_i
  ↓
  s

/* s_total = sum{s_i} */
```

Author: incorporated into the writeup, thanks!
